Added optional GDCM support to enable reading of compressed pixel data.

Korijn opened the pull request:
This PR adds logic to parse pixel data using GDCM if it can be imported. Tested on compressed MR, CT and X-ray data, using GDCM 2.6.1 and pydicom HEAD.

Note: `conda` users can add the `clinicalgraphics` channel to easily install the `gdcm` package.

Korijn commented:
This should resolve #18.

cancan101 commented:
Does this properly handle DICOMs that are compressed with LUTs (look up tables)?

For reference help for `gdcmconv`:
```
Image only Options:
  -l --apply-lut                      Apply LUT (non-standard, advanced user only).
```

Korijn commented:
This PR only reads the image data into a numpy array, and subsequently leaves it to pydicom to unpack the data correctly.

If you can point me to an example file I can try it and see what happens.

NB.: I know the code could be cleaner but I wanted to see everyone's initial responses first.

cancan101 commented:
I think you want to have GDCM decode the LUTs and return the actual pixel values.

Korijn commented:
Always or optionally?

Is this really important for this PR or can it be done separately? You are welcome to fork my PR and send me a PR for the PR ;)

cancan101 commented:
I don't think this needs to be solve here, but it is worth making sure the
fact that a LUT is in use isn't discarded and just mapped to a monochrome
image.

On Wed, Dec 2, 2015 at 3:12 PM Korijn van Golen <notifications@github.com>
wrote:

> Always or optionally?
>
> Is this really important for this PR or can it be done separately? You are
> welcome to fork my PR and send me a PR for the PR ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/darcymason/pydicom/pull/232#issuecomment-161420075>.
>

Korijn commented:
Ok, I'll check it out.

darcymason commented:
This is very nice ... I'll hold off on pulling it in until you look at the LUT issue. 

I added a couple of minor comments to the changeset, if you are making further edits.

Another suggestion:  unit tests for this would be ideal -- there are a few compressed format images in the testfiles directory that could be tested.  If you don't have time to code the tests I can add an issue for that.  The checks should of course be skipped if gdcm is not installed.

cancan101 commented:
Perhaps add a note in the readme on how to install the `gdcm `package on various OS' (Windows, Linux, OS X)? It looks like the conda solution works for only Windows.

Korijn commented:
I updated the PR with cleaned up code. I followed up on @darcymason's comments, and, @cancan101: I also added LUT support (but only for the GDCM fallback...).

@cancan101 I'll see if I can add some explanatory text for GDCM installation, although it's not really trivial. It took me a week to get the conda packages for all Python versions and CPU architectures out there...

cancan101 commented:
I am guessing that `Unsupported JPEG data precision 12` means the file has 12 bit precision but it get's down cast to 8bit?

cancan101 commented:
>I was unable to test for the LUT support though, because that only happens in the case that GDCM is used and that's only when pydicom can't handle the pixel data natively. That case is very rare I believe (JPEG compression + PALETTE_COLOR).
If you look at #205, currently pydicom thinks it can handle `PALETTE_COLOR` but it actually does not correctly. You probably want to use gcdm even if the file isn't compressed in the case of `PALETTE_COLOR`

cancan101 commented:
This is certainly beyond the scope of this PR, but it would be nice to see see a pip package for gcdm.

Korijn commented:
Well, I just parsed the LUT DICOM files correctly using GDCM, and fixed some more errors. Using GDCM first now yields:

```

- tests\test_files\nested_priv_SQ.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\OT-PAL-8-face.dcm
  Success

- tests\test_files\OBXXXX1A.dcm
  Success

- tests\test_files\rtplan_truncated.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\ExplVR_BigEndNoMeta.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\rtdose.dcm
  Success

- tests\test_files\ExplVR_LitEndNoMeta.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\CT_small.dcm
  Success

- tests\test_files\priv_SQ.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\JPEG-lossy.dcm
  Success

- tests\test_files\JPEG-LL.dcm
  Success

- tests\test_files\test-SR.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\image_dfl.dcm
  Failure: GDCM needs the filename to read.

- tests\test_files\JPEG2000.dcm
  Success

- tests\test_files\reportsi.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\color-pl.dcm
  Success

- tests\test_files\no_meta_group_length.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\MR_small.dcm
  Success

- tests\test_files\rtplan.dcm
  Failure: GDCM could not read DICOM image

- tests\test_files\color-px.dcm
  Success

- tests\test_files\ExplVR_BigEnd.dcm
  Success

- tests\test_files\rtstruct.dcm
  Failure: GDCM could not read DICOM image
```

...which is probably the best it can do! :)

I'm still wondering if we should use GDCM as a fallback or primarily if it is available. It does fix the LUT issue if you use it as the primary approach.

Korijn commented on pydicom/dataset.py:
```diff
@@ -355,78 +362,164 @@ def __iter__(self):
+        pixel_array = None
+        
+        # FIXME: should we always use GDCM if it is available? Or only as a fallback?
+        if self._is_supported_transfer_syntax(): # and not have_gdcm
```
If we uncomment the end of line 392, GDCM will be used first if it is available. If we leave it commented out, it will only be used as a fallback.

Main benefit of using GDCM first: GDCM supports LUT's, pydicom does not.

Korijn commented on pydicom/dataset.py:
```diff
@@ -355,78 +362,164 @@ def __iter__(self):
+                
+                # amount of samples per pixel can change
+                self.SamplesPerPixel = gdcm_lutfiltered_pixmap.GetBufferLength() / gdcm_buf_len
+                self.PlanarConfiguration = 0
```
I don't like how I overwrite `self.SamplesPerPixel` and `self.PlanarConfiguration` in the case of a LUT. It feels like I might be corrupting internal state.

Do you guys have any suggestions to handle this more elegantly?

Korijn commented on pydicom/dataset.py:
```diff
@@ -355,78 +362,164 @@ def __iter__(self):
+                        try:
+                            gdcm_raw_data_buffer = encode_gdcm_buffer_as_bytearray(gdcm_raw_data_buffer, encoding='mbcs')
+                        except UnicodeEncodeError:
+                            raise TypeError("Could not encode the image unicode string returned by GDCM back to a byte array")
```
Obviously I should do this in a loop over possible encodings, but even better would be if I could somehow determine the correct encoding intelligently. Any suggestions?

Korijn replied:
Also, which other encodings should I include?

cancan101 replied:
What is the rationale for iterating over / trying different encodings?

Korijn replied:
Short answer: it's complicated.

Long answer: GDCM returns the byte buffer as `char*` in c++. This is returned as a `str` by the Python wrappers generated by SWIG. 

On Python 2 `str` is a byte buffer by default and everything is OK. On Python 3, `str` is a unicode string and no longer a byte buffer. Since we want to access the bytes directly we need to re-encode the unicode string to their original corresponding bytes (a `bytearray` in py3).

This requires us to know the original encoding of the file, which we can only guess, as far as I am aware. This is why we need to try a bunch of codecs. Ideally we would be able to slecet the appropriate codec directly, but I don't know how to do that.

We also need to specify the special surrogate error handler since some bytes have no corresponding unicode character and this would result in an exception under normal conditions.

Also see the SWIG docs on this topic:
http://www.swig.org/Doc3.0/Python.html#Python_nn77

This article also explains a lot about unicode handling in py3:
http://python-notes.curiousefficiency.org/en/latest/python3/text_file_processing.html

cancan101 replied:
This part sounds odd to me:
>This requires us to know the original encoding of the file, which we can only guess, as far as I am aware. This is why we need to try a bunch of codecs. Ideally we would be able to slecet the appropriate codec directly, but I don't know how to do that.

>We also need to specify the special surrogate error handler since some bytes have no corresponding unicode character and this would result in an exception under normal conditions.

Is there no way to get the swagger wrapper to return a byte array on python3?

cancan101 replied:
I guess I understand the need for `surrogateescape` but not for guessing the codec.

Korijn replied:
If you check out the link ti the SWIG docs on this topic you'll find that they think it's not their problem.

Korijn replied:
@malaterre do you have any suggestions? Should the SWIG bindings in GDCM be adjusted perhaps?

cancan101 replied:
Reading that link, it seems like it should just be encoded as `'utf-8'`. I don't see it saying  to try other encodings.

Korijn replied:
I'll give it a shot.

cancan101 commented on pydicom/dataset.py:
```diff
@@ -355,78 +362,164 @@ def __iter__(self):
+            gdcm_raw_data_buffer = None
+            
+            # in the case of a lookup table try to apply it
+            if gdcm_image.GetPhotometricInterpretation().GetType() == gdcm.PhotometricInterpretation.PALETTE_COLOR:
```
Can the code to handle `PALETTE_COLOR` be factored out as its own function?

Korijn replied:
Yes, but it isn't usable without GDCM, and it has some side effects such as changing the `SamplesPerPixel` property. I'm not sure how to handle that correctly (@darcymason ?). Also see my other note

cancan101 commented:
I would add a note in the README that input needs to be an actual file rather than a stream. This can be problematic in case where you are receiving the content using pynetdicom and explicitly don't want the data to be saved to disk.

cancan101 commented:
That being said, it looks like there _might_ be some support for reading from streams / files, I don't know how this gets SWIGed: http://gdcm.sourceforge.net/html/classgdcm_1_1Reader.html#a2fdbf80e6dedc1f3a127a68d6ab11341 and http://gdcm.sourceforge.net/html/classgdcm_1_1Reader.html#a970fc96235a832e27eef6d1d0a422765

malaterre commented:
I am open for *any* patches in GDCM that can improve the swig python wrapping. I have mostly been working with python 2.x, C# and java these days.

malaterre commented:
As a side note, GDCM does not support MPEG2 or MPEG4 (or even the upcoming MPEG5).

malaterre commented:
As a side note, if you want streaming. You may want to check gdcm::ImageRegionReader. See for example http://gdcm.sourceforge.net/html/ExtractImageRegion_8cs-example.html

Korijn commented:
@malaterre thanks for your input. Ideally I'd only pass the `self.PixelData` `bytearray` and possibly some other required parameters and get the decompressed `PixelData` back from GDCM, to avoid using the `self.filename` which is not available in the case of processing a file stream directly.

Korijn commented:
Updated the code:

* Added FIXMEs for future improvements
* Removed the LUT application as it is beyond the scope of this PR (and the time I have available)

I wouldn't mind if the PR would be merged in it's current state.

If @malaterre can shed some light on how we can use the `self.PixelData` property directly instead of `self.filename`, I might be able to resolve both FIXME's. Otherwise, no pie.

cancan101 commented on pydicom/dataset.py:
```diff
@@ -355,78 +362,162 @@ def __iter__(self):
+            # if we use self.PixelData directly as the FIXME above indicates this will no longer be an issue
+            if sys.version_info >= (3, 0):
+                py3_encoding_succeeded = False
+                for encoding in ["utf-8", "mbcs", sys.getfilesystemencoding()]:
```
Per my comments, I don't see why using `"utf-8"`  is not enough?

Korijn replied:
It took me a while to realize that you are correct. I'll update that next Monday.

Korijn commented:
Fix unicode encoding technique (thanks @cancan101), and cleaned up the code significantly.

I don't see how we can improve this any further without someone helping me figure out how to pass `self.PixelData

darcymason commented:
Thanks again, Korijn, this is a great contribution.  I have merged the pull request.

We still have an outstanding issue I think on how to deal with LUTs, especially if not using GDCM.  I did some reading on that and it seems quite complex, with possible multiple LUTs being applied.  There is also the complication then of how to write such files, particularly if the user has modified the pixel data.  Needs some more thought.  But perhaps that is a discussion for #205.

Korijn commented:
My pleasure, I can't wait to implement our DICOM handling code without verbose C wrappers...

I guess I should have included this in the PR in the first place, but maybe we should add a note to the README.md and the docs about the new GDCM support for compressed image data...
